components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/Keystore.kt

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.dataprotect
+
+import android.annotation.TargetApi
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyProperties
+import java.security.KeyStore
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+import javax.crypto.spec.GCMParameterSpec
+
+private val KEYSTORE_PROVIDER = "AndroidKeyStore"
+private val ENCRYPTED_VERSION = 0x02
+
+private val CIPHER_ALG = KeyProperties.KEY_ALGORITHM_AES
+private val CIPHER_MOD = KeyProperties.BLOCK_MODE_GCM
+private val CIPHER_PAD = KeyProperties.ENCRYPTION_PADDING_NONE
+private val CIPHER_TAG_LEN = 128
+private val CIPHER_SPEC = "${CIPHER_ALG}/${CIPHER_MOD}/${CIPHER_PAD}"
```

Comment:
Are we expecting to pick values here that will work for every app or do you think this should be configurable by a consuming app (constructor parameters with default values? e.g. `val cipherMod: String = KeyProperties.BLOCK_MODE_GCM`)?

Reply:
I think there's little-to-no value in making these configurable; this crypto is just on the device, for the app consuming the API.  It (the key, therefore the encrypted data) can't leave the device.

The algorithm spec chosen is the strongest available, and I think it should be enforced.

Reply:
👍